### PR TITLE
Introduce error margin by variable in tests

### PR DESCRIPTION
**Pull request by benjello**
Thanks for contributing to OpenFisca! Remove this line, as well as any other in the following that don't fit your contribution  :)

#### New features

- Introduce variable dependent error margins int YAML tests
  - Allows for different error margin across variables

### CODE REVIEW

**Review by nikhilwoodruff — APPROVED** [review]
Great, thanks @benjello. This will be useful in both UK and US systems.

**Comment by benjello** [comment]
Doc building fails and should be fixed by #1145.

@maukoquiroga and @MattiSG please ping me when the later is merged so I can move on with this PR.
Thanks !

**Comment by maukoquiroga** [comment]
Hi @benjello ! Thanks for your contribution. Could you please write a couple of lines so I or an other reviewers can understand the motivation for this PR? Just 

1. Need
2. Complication to solve the need with current code base
3. What you propose and why you propose it this way and not another way

Quickly from reading the code this is what I got:

1. We already accept calculation level error margins, which are absolute. 
2. However, because quality of data is not always homogeneous, in some cases, outlier errors in one variable calculation makes the whole calculation to fail. Which is generally OK except when we want to observe one variable in isolation.
3. You propose to define 
  a. A calculation-level error margin that is called absolute in terms of the desired result.
  b. A variable-level error margin that is relative to the absolute one and the desired result.
  c. The `default` error is absolute and is a scalar, e.g. `100 €`.
  d. The `relative` error is relative to the `default` and is a percentage, e.g. `5%`.

File: openfisca_core/tools/test_runner.py
Comment (maukoquiroga) [inline openfisca_core/tools/test_runner.py:218]: Why call it default and not just absolute error margin?

File: openfisca_core/tools/test_runner.py
Comment (maukoquiroga) [inline openfisca_core/tools/test_runner.py:219]: Can't we rather assume that if not declared, relative error margin is 0%?

File: openfisca_core/tools/test_runner.py
Comment (maukoquiroga) [inline openfisca_core/tools/test_runner.py:237]: How did this work previously, do you have an example?

File: tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml
Comment (maukoquiroga) [inline tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml:1]: I'm not clear here: the absolute error margin is for the whole calculation or just for the variable?

File: tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml
Comment (maukoquiroga) [inline tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml:5]: Ah, so here you declare an absolute for income tax overriding the abdolute of the calculaiton ?

File: tests/fixtures/yaml_tests/failing_test_variable_relative_error_margin.yaml
Comment (maukoquiroga) [inline tests/fixtures/yaml_tests/failing_test_variable_relative_error_margin.yaml:4]: I don't understand here the logic. Relative error margin to relative to what? Are we here again just overriding the relative margin error just fro income tax? 

Maybe an example where this is useful including other variables could come handy to better grasp the intent.

File: tests/fixtures/yaml_tests/test_relative_error_margin.yaml
Comment (maukoquiroga) [inline tests/fixtures/yaml_tests/test_relative_error_margin.yaml:14]: So here we demonstrate that test passes because the result is within the lower and upper bounds (+-15), but should otherwise fail with the general one (+-3). Could we show that in the test to it's clear?

**Comment by maukoquiroga** [comment]
This goes beyond the PR because error margin logic already exist in OF, but to be clear, when we say _income tax has a relative margin error of 5%_, we're saying just that _income tax can have an absolute margin error of `+- result * relative error margin`_ right?

File: openfisca_core/tools/test_runner.py
Comment (benjello) [reply openfisca_core/tools/test_runner.py:219]: You want to be able to easily (continuously) deviate from the current situation where you have a global error margin

File: openfisca_core/tools/test_runner.py
Comment (benjello) [reply openfisca_core/tools/test_runner.py:237]: Current tests clearly shows how margins can be used today.

File: tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml
Comment (benjello) [reply tests/fixtures/yaml_tests/failing_test_variable_absolute_error_margin.yaml:1]: The default is for all the variables that are not directly specified.
Imagine you set a 1% relative error margin. 
```yaml
relative_error_margin: .01
```
But you find out that want to have a less strict, say 5%, error margin on some variables.
You can easily modify the previous line to:
```yaml
relative_error_margin: 
- default: .01
- my_variable: .05
```

File: tests/fixtures/yaml_tests/failing_test_variable_relative_error_margin.yaml
Comment (benjello) [reply tests/fixtures/yaml_tests/failing_test_variable_relative_error_margin.yaml:4]: relative error margin = abs(computed_value - target_value / target_value)  is here 100%.

An example is pension there are some variables where it is very difficult to compute up to 1% error.
And other where it is crucial.

**Comment by benjello** [comment]
> This goes beyond the PR because error margin logic already exist in OF, but to be clear, when we say _income tax has a relative margin error of 5%_, we're saying just that _income tax can have an absolute margin error of `+- result * relative error margin`_ right?

See above for the definition of relative error margin

**Comment by benjello** [comment]
@maukoquiroga : does my answer clarify the PR. May be we can make it clearer in the doc after merging the PR.

**Comment by maukoquiroga** [comment]
> @maukoquiroga : does my answer clarify the PR. May be we can make it clearer in the doc after merging the PR.

Yes, I understand better.

The reason I ask all this questions is to make it easier for any reviewer out there to get the logic and review the code, and eventually accept/reject the proposal.

I certainly have no grounds to reject it now that I understand the value it adds, however I wasn't able to review the actual implementation of it without that understanding.

I hope to give you a proper review soon if nobody gives you one before.

File: openfisca_core/tools/test_runner.py
Comment (benjello) [reply openfisca_core/tools/test_runner.py:218]: It is the default of the absolute error margin

**Comment by benjello** [comment]
@nikhilwoodruff @maukoquiroga : I do need a final approval to merge this PR ;-)

**Review by maukoquiroga — APPROVED** [review]

**Pull request merged**
